### PR TITLE
CFY-4988 Add blueprint diff

### PR DESCRIPTION
**Pull request by AviaE**

### CODE REVIEW

File: dsl_parser/blueprint_diff.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+#    * limitations under the License.
+
+
+def blueprint_diff(old_blueprint, new_blueprint):
+
+    chain = []
+    output = {}
+
+    diff(old_blueprint, new_blueprint, chain, output, 'added')
+    diff(new_blueprint, old_blueprint, chain, output, 'removed',
```
Comment (AviaE) [inline dsl_parser/blueprint_diff.py]: use the mode argument to decide if to add 'modified'

File: dsl_parser/blueprint_diff.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+         include_modified=False)
+
+    return output
+
+
+def diff(old, new, chain, output, mode, include_modified=True):
```
Comment (AviaE) [inline dsl_parser/blueprint_diff.py]: change chain to 'path' or something similar

that also reflects on the `key` var in the main loop

File: dsl_parser/blueprint_diff.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+        value = new[key]
+        chain.append(key)
+
+        if not has_chain(old, chain):
+            add_chain(output, chain, mode)
+
+        elif type(value) is not dict:
+            old_value = get_value(old, chain)
+            if value != old_value and include_modified:
```
Comment (AviaE) [inline dsl_parser/blueprint_diff.py]: check if the == is deep comparison

File: dsl_parser/blueprint_diff.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+                if type(value) is list:
+                    modified = list_diff(old_value, value)
+                else:
+                    modified = 'modified'
+                add_chain(output, chain, modified)
+
+        else:
+            diff(old, value, chain, output, mode, include_modified)
+
+        del chain[-1]
```
Comment (AviaE) [inline dsl_parser/blueprint_diff.py]: check if we can delete the last element in a prettier way

File: dsl_parser/blueprint_diff.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+
+def list_diff(old, new):
+
+    added_items = set(range(len(new)))
+    output = {}
+    for idx, item in enumerate(old):
+
+        # item was removed
+        if item not in new:
```
Comment (AviaE) [inline dsl_parser/blueprint_diff.py]: again, does that include deep comparison?

File: dsl_parser/blueprint_diff.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        for key in chain:
+            d = d[key]
+
+    except KeyError:
+        return False
+
+    return True
+
+
+def add_chain(output, chain, value):
```
Comment (AviaE) [inline dsl_parser/blueprint_diff.py]: change to `add_entry` or something

File: dsl_parser/tests/test_blueprint_diff.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+
+class DiffTestCase(unittest.TestCase):
+
+    def test_convert_dict_to_chain(self):
+        empty_chain = []
+        chain_with_one_key = ['a']
+        long_chain = ['a', 'b', 'c']
+
+        value = 'added'
```
Comment (AviaE) [inline dsl_parser/tests/test_blueprint_diff.py]: delete this and use 'stub_value'

File: dsl_parser/tests/test_blueprint_diff.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+    @patch('dsl_parser.blueprint_diff.convert_chain_to_dict')
+    def test_add_chain_to_output(self, convert_chain_mock):
+
+        output = {'a': {'b': {'c': 3}, 'd': {'e': 5}}, 'f': 6}
+        chain = ['a', 'b', 'h']
+        add_chain(output, chain, stub_value)
+
+        convert_chain_mock.assert_called_with(chain[2:], stub_value)
+
```
Comment (AviaE) [inline dsl_parser/tests/test_blueprint_diff.py]: Also add a test that doesn't mock convert_chain

File: dsl_parser/tests/test_blueprint_diff.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    def test_has_chain(self):
+
+        output = {'a': {'b': {'c': 3}, 'd': {'e': 5}}, 'f': 6}
+
+        chain_a = ['a', 'b', 'd']
```
Comment (AviaE) [inline dsl_parser/tests/test_blueprint_diff.py]: change that names to more meaningful ones.
like:
non existent
exists
deep exists

File: dsl_parser/tests/test_blueprint_diff.py [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+        chain_a = ['a', 'b', 'c']
+        chain_b = ['a', 'd', 'e']
+        chain_c = ['f']
+
+        self.assertEqual(get_value(output, chain_a), 3)
+        self.assertEqual(get_value(output, chain_b), [4, 5])
+        self.assertEqual(get_value(output, chain_c), '6')
+
+    def assert_diff_result(self, diff_function,
```
Comment (AviaE) [inline dsl_parser/tests/test_blueprint_diff.py]: underscore

File: dsl_parser/blueprint_diff.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+
+def diff(old, new, path_of_keys, output, mode):
+
+    for key in new:
+
+        value = new[key]
+        path_of_keys.append(key)
+
+        if not has_path_of_keys(old, path_of_keys):
```
Comment (ChenRoth) [inline dsl_parser/blueprint_diff.py]: rename 'has_path_of_keys' to simply 'path_exists'

File: dsl_parser/blueprint_diff.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+    path_of_keys = []
+    output = {}
+
+    diff(old_blueprint, new_blueprint, path_of_keys, output, 'added')
+    diff(new_blueprint, old_blueprint, path_of_keys, output, 'removed')
+
+    return output
+
+
+def diff(old, new, path_of_keys, output, mode):
```
Comment (ChenRoth) [inline dsl_parser/blueprint_diff.py]: if mode is binary, it's best to just be a boolean argument with a semantic change

File: dsl_parser/blueprint_diff.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+                    modified = 'modified'
+                add_entry(output, path_of_keys, modified)
+
+        else:
+            diff(old, value, path_of_keys, output, mode)
+
+        del path_of_keys[-1]
+
+
+def list_diff(old, new):
```
Comment (ChenRoth) [inline dsl_parser/blueprint_diff.py]: it'd be helpful to add a docstring describing how list diffs are treated

Comment (AviaE) [inline dsl_parser/blueprint_diff.py]: add example

**Comment by AviaE** [comment]
reviewed by @ChenRoth

**Pull request merged**